File: leetcode_2466/CountWaysToBuildGoodStrings.java
package leetcode_2466;

import java.util.Arrays;

class CountWaysToBuildGoodStrings {
    private static final int MOD = 1000000007;
    private int [] dp;

    // Top-down approach 
    // public int countGoodStrings(int low, int high, int zero, int one) {
    //     dp = new int[high + 1];
    //     Arrays.fill(dp, -1);
        
    //     return dfs(0, low, high, zero, one);

    // }

    // private int dfs(int length, int low, int high, int zero, int one) {
    //     if(length > high)
    //         return 0;
        
    //     if(dp[length] != -1)
    //         return dp[length];
        
    //     int count = length >= low ? 1 : 0;
    //     count = (count + dfs(length + zero, low, high, zero, one)) % MOD;
    //     count = (count + dfs(length + one, low, high, zero, one)) % MOD;

    //     dp[length] = count;
    //     return dp[length];
    // }

    // Bottom-up approach
    public int countGoodStrings(int low, int high, int zero, int one) {
        dp = new int [high + 1];
        dp[0] = 1; // Base case: there is only one way to build a string of length 0

        for(int i = 1; i <= high; i++)
        {
            if(i >= zero)
                dp[i] = (dp[i] + dp[i - zero]) % MOD;
            if (i >= one)
                dp[i] = (dp[i] + dp[i - one]) % MOD;
        }

        int result = 0;
        for(int i = low; i <= high; i++)
            result = (result + dp[i]) % MOD;
        
        return result;
    }

    public static void main(String[] args) {
        CountWaysToBuildGoodStrings sol = new CountWaysToBuildGoodStrings();
        int low = 3, high = 3, zero = 1, one = 1;
        System.out.println(sol.countGoodStrings(low, high, zero, one));
    }
}

File: leetcode_2466/CountWaysToBuildGoodStrings.py
class CountWaysToBuildGoodStrings:
    def countGoodStrings(self, low: int, high: int, zero: int, one: int) -> int:
        # Bottom up DP Approach
        dp = {0 : 1}
        mod = 10**9 + 7
        for i in range(1, high + 1):
            dp[i] = (dp.get(i - zero, 0) + dp.get(i - one, 0)) % mod
        
        return sum([dp[i] for i in range(low, high + 1)]) % mod;
    
        # # Top down DP Approach
        # MOD = 10**9 + 7
        # dp = {}
        
        # def dfs(length):
        #     if length > high:
        #         return 0

        #     if length in dp:
        #         return dp[length] 
            
        #     dp[length] = 1 if length >= low else 0
        #     dp[length] += dfs(length + zero) + dfs(length + one)
            
        #     return dp[length] % MOD
    
        # return dfs(0)
    
def main():
    countWaysToBuildGoodStrings = CountWaysToBuildGoodStrings()
    print(countWaysToBuildGoodStrings.countGoodStrings(3, 3, 1, 1))

if __name__ == "__main__":
    main()
    
    